### PR TITLE
Reduce file watcher memory usage

### PR DESCRIPTION
**Pull request by pvid**
This PR replaces `FileTreeRepository` with `PathWatcher`. The former keeps a cached version of watched file trees while the latter does not. Since the implementation no longer hashes files, but looks only on modification time, we can only watch modifications without keeping any state.

This makes a huge difference especially on macOS, where in large projects the root of the project has to be watched. My motivation for this change is that from time to time, metals stopped working and threw OOM exceptions (I think I experienced something akin to https://github.com/scalameta/metals/issues/3020) and I had to delete the `.bloop` folder and restart metals. In a particular case, the cached paths took more than half of 4GB heap dedicated to metals.

Shoutout and huge thanks to @eatkins that fixed an issue https://github.com/swoval/swoval/issues/155 that I experienced while testing the change on Linux.

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/TargetData.scala
Comment (pvid) [inline metals/src/main/scala/scala/meta/internal/metals/TargetData.scala:54]: not too happy with creating another random set here. I was thinking of keeping some more information about source items, maybe in its original BSP form as `Map[AbsolutePath, SourceItem]` that could also be used for distinguishing generated dirs



This version has the advantage of being very straight forward and since the vast majority of source items are directories, the set is very small

what do you think?

**Comment by eatkins** [comment]
@pvid I am wondering if I pushed an update to swoval that allowed the FileTreeRepository to filter out the paths you don't care about if that would be a simpler fix? If you want to try a branch that should handle the use case, I pushed it here: https://github.com/swoval/swoval/pull/158. I think the fix with the swoval update might be as simple as:
```diff
- val repo = FileTreeRepositories.get(converter, /*follow symlinks*/ true)
+ val repo = FileTreeRepositories.get(converter, tp => watchFilter(tp.getPath), /*follow symlinks*/ true)
```
but I haven't tried it myself.

**Comment by pvid** [comment]
> @pvid I am wondering if I pushed an update to swoval that allowed the FileTreeRepository to filter out the paths you don't care about if that would be a simpler fix? If you want to try a branch that should handle the use case, I pushed it here: [swoval/swoval#158](https://github.com/swoval/swoval/pull/158). I think the fix with the swoval update might be as simple as:
> 
> ```diff
> - val repo = FileTreeRepositories.get(converter, /*follow symlinks*/ true)
> + val repo = FileTreeRepositories.get(converter, tp => watchFilter(tp.getPath), /*follow symlinks*/ true)
> ```
> 
> but I haven't tried it myself.

that would certainly help with the problem I described, but conceptually, there is still no reason to keep a cached representation given metals uses just the PathWatcher functionality

**Comment by eatkins** [comment]
Fair enough. One drawback to PathWatcher is that it isn't very reliable in terms of correctly identifying what kind of event happened. For instance, it isn't possible to distinguish a file creation from a file modification on macOS. In order to get the event types right, it is necessary to cache the state of every file so the additional overhead of storing say a last modified time or hash is small. But it sounds like getting the event types right may not really matter for metals in which case PathWatcher is fine.

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
Comment (eatkins) [inline metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:182]: It isn't maybe idiomatic scala, but the way I like to do something like this is:
```scala
val thread = new Thread("metals-watch-callback-thread") {
  @tailrec def loop(): Unit = {
    try callback(watchEventQueue.take)
    catch { case _: InterruptedException =>  }
    if (!stopWatchingSignal.get) loop()
  }
  override def run(): Unit = loop()
}
...
() => {
  watcher.close()
  stopWatchingSignal.set(true)
  thread.interrupt()
}
```

It's a minor thing but with this technique the watch thread stays idle until there really is work to do and it will also shutdown essentially instantaneously instead of on average taking 100ms.

**Comment by pvid** [comment]
> Fair enough. One drawback to PathWatcher is that it isn't very reliable in terms of correctly identifying what kind of event happened. For instance, it isn't possible to distinguish a file creation from a file modification on macOS. In order to get the event types right, it is necessary to cache the state of every file so the additional overhead of storing say a last modified time or hash is small. But it sounds like getting the event types right may not really matter for metals in which case PathWatcher is fine.

that is a good point with the type of events. Since metals does the same thing for CREATE and MODIFY, I will coalesce them into a single event type here with a comment about the shortcomings of `PathWatcher` and that `FileTreeRepositories` should be used when that distinction is important.

I will leave it to the maintainers to decide if they want to stick with `FileTreeRepositories` or not. Thank you for your input! I really does better illuminate the pros and cons of both

**Comment by eatkins** [comment]
Happy to help. I just want to add for any reviewers that internally a FileTreeRepository delegates to a PathWatcher so this should be a safe change provided that the distinction between creation and modification events is not important.

File: metals/src/main/scala/scala/meta/internal/metals/TargetData.scala
Comment (tanishiking) [reply metals/src/main/scala/scala/meta/internal/metals/TargetData.scala:54]: > I was thinking of keeping some more information about source items, maybe in its original BSP form as Map[AbsolutePath, SourceItem] that could also be used for distinguishing generated dirs

I was considering adding those additional data (like `SourceItem`) to `buildTargetSources` :)
https://github.com/scalameta/metals/blob/9fe87ba4bd2abaf41a352e4cb7146df45535b78d/metals/src/main/scala/scala/meta/internal/metals/TargetData.scala#L41-L42

While creating another random set here wouldn't be a big problem, can we stay `path.isScalaOrJava` here https://github.com/scalameta/metals/pull/3758/files#diff-314e34520d936536cd836f653612edc262cf2afb106ae5072025978256a1fa8cL83 ? As the argument of the `collect` method is from `buildTarget.sourceItems`, they are obviously a source item of the build target.

Or can we use `buildTargetSources`? (By somehow giving a pair of `AbsolutePath` and `BuildTargetIdentifier` to the `collect` method).

**Review by tanishiking — APPROVED** [review]
I haven't yet tested this branch locally, but the change looks on the right track, great work! I left some comments.

> provided that the distinction between creation and modification events is not important

In terms of a distinction between creation and modification, it should be safe in Metals as we invoke the same callback method for both creation and modification.
https://github.com/scalameta/metals/blob/9fe87ba4bd2abaf41a352e4cb7146df45535b78d/metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala#L1375-L1378

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
Comment (tanishiking) [reply metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:182]: I prefer this version :) `watchEventQueue.take` will block until there's an event in the queue and the watch thread stays idle 😍

**Review by tanishiking — COMMENTED** [review]
Sorry, I approved by mistake (while it looks good:+1: I want to test it locally), I was just commenting

**Comment by kpodsiad** [comment]
> This makes a huge difference especially on macOS, where in large projects the root of the project has to be watched. [...] In a particular case, the cached paths took more than half of 4GB heap dedicated to metals.

@pvid is this >2GB of cached paths after https://github.com/scalameta/metals/pull/3665 or before?

I'm good with using `PathWatcher`. If at some point of the time it would be necessary to distinguish between event types we could always rollback to the `FileTreeRepository`.

**Comment by pvid** [comment]
> > This makes a huge difference especially on macOS, where in large projects the root of the project has to be watched. [...] In a particular case, the cached paths took more than half of 4GB heap dedicated to metals.
> 
> @pvid is this >2GB of cached paths after #3665 or before?
> 
> I'm good with using `PathWatcher`. If at some point of the time it would be necessary to distinguish between event types we could always rollback to the `FileTreeRepository`.

that PR is useful for monorepos with just a few scala modules (see default limit of 32 roots). The project I am working on has > 100 modulesand it would require more roots than macOS file watching is able to handle.

File: metals/src/main/scala/scala/meta/internal/metals/TargetData.scala
Comment (pvid) [reply metals/src/main/scala/scala/meta/internal/metals/TargetData.scala:54]: The reason for going from `path.isScalaOrJava` to getting info from BSP is that some tests were failing, because of a `*.sbt` source file file that failed that check. However, it still needs to be watched as a file.

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
Comment (pvid) [reply metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:182]: changed in a fixup commit

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
Comment (pvid) [reply metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:144]: my mistake, I changed the value for some testing. Fixed in a fixup commit

**Review by tanishiking — APPROVED** [review]
LGTM from my side, do you have any additional comments? @kpodsiad

**Comment by pvid** [comment]
I have squashed the fixup commits resulting from the review

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:79]: I have some proposition of refactor for `collect`
```scala
def collect(path: AbsolutePath): Unit = {
  val shouldBeWatched =
    // Do nothing, already covered by a source root
    !buildTargets.isInsideSourceRoot(path) &&
      !buildTargets.checkIfGeneratedSource(path.toNIO)

  if (shouldBeWatched) {
    if (buildTargets.isSourceFile(path)) files.add(path.toNIO)
    else directories.add(path.toNIO)
  }
}
```

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:209]: We should also print the stack trace of the encountered error.

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcherEvent.scala
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcherEvent.scala:27]: What exactly indicates `EventType.Overflow`? I think it'll be worth adding some comment next to the `case object Overflow extends EventType`

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:218]: Could you remove the brackets surrounding the cases' bodies?

**Review by kpodsiad — COMMENTED** [review]
Looks good, I have only a few suggestions. 
Also, I'm going to test this change manually today.

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcherEvent.scala
Comment (pvid) [inline metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcherEvent.scala:14]: TODO merge Create and Modify events, see https://github.com/scalameta/metals/pull/3758#issuecomment-1080005296

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcherEvent.scala
Comment (pvid) [reply metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcherEvent.scala:27]: Overflow indicates that the file watcher was overwhelmed and may have lost some events . I will add a comment

File: metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
Comment (pvid) [reply metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala:218]: > Looks good, I have only a few suggestions. Also, I'm going to test this change manually today.

Thanks! I will address your comments this evening. I also noticed that I forgot to address one issue, see https://github.com/scalameta/metals/pull/3758#discussion_r841469303

**Comment by kpodsiad** [comment]
@pvid Will it be a problem if this PR got merged after a Metals release (which may happen this week)? 
This PR touches a Metals core so it would be safer to merge this after the stable version is released. Thanks to that we will have more time to battle-test this until the next stable release. 
So, will it be ok for you and your colleagues at work to use the snapshot version with these changes?

**Comment by tgodzik** [comment]
I need to still look into some possible issues with Bloop before the next release, so I would rather postpone it if possible.

**Comment by kpodsiad** [comment]
Don't worry @tgodzik, with @dos65 we are aware of some of these issues and there won't be any release without resolving them first. Enjoy the rest of your holidays, I'm not cutting a release without you knowing about it :D

**Comment by pvid** [comment]
@kpodsiad sure, no problem

**Comment by pvid** [comment]
I have rebased the changes onto current main and addressed all the comments and suggestions from review. It should be ready for merge after the next release as noted above.

I will be using a SNAPSHOT version of metals with these changes to make sure that some unforeseen issues do not arise in day-to-day use

**Comment by tanishiking** [comment]
Now the latest version is out, it's time to merge 🚀

**Pull request merged**
